continuationsPlugin/src/main/scala/continuations/SafeContinuation.scala

```diff
@@ -29,27 +28,20 @@ class SafeContinuation[-T](val delegate: Continuation[T], initialResult: Any | N
   def getOrThrow(): Any | Null | Continuation.State.Suspended.type =
     var result = this.result
     if (result == Continuation.State.Undecided)
-      if (SafeContinuation
-          .RESULT
-          .compareAndSet(this, Continuation.State.Undecided, Continuation.State.Suspended))
+      if (CAS_RESULT(Continuation.State.Undecided, Continuation.State.Suspended))
         return Continuation.State.Suspended
       result = this.result
     if (result == Continuation.State.Resumed) Continuation.State.Suspended
     else if (result.isInstanceOf[Left[_, _]])
       throw result.asInstanceOf[Left[Throwable, _]].value
-    else result
+    else if (result.isInstanceOf[Right[_, _]])
+      result.asInstanceOf[Right[_, _]].value
+    else
+      result // COROUTINE_SUSPENDED
```

Comment:
I believe this is right, as mentioned [here](https://github.com/JetBrains/kotlin/blob/master/libraries/stdlib/jvm/src/kotlin/coroutines/SafeContinuationJvm.kt#L59).

It is not quite clear to me whether this `COROUTINE_SUSPENDED` would be inside the `Right`, in which case this last `else` should throw something like `throw IllegalStateException("Unexpected result value" ++ result.toString)` or the `COROUTINE_SUSPENDED` would be the value of the result.

In either case I think it is fine to return the `result`, and re-evaluate when we have build all the functionality that handles the response from this `getOrThrow` method

Reply:
> or the COROUTINE_SUSPENDED would be the actual value of the result.

It's the actual value of the result. In our case `Continuation.State.Suspended`. This can be coming from other surrounding suspending code, and needs to be passed through. It signals to the CPS machinery that at some later point `resumeWith` will be called, which will in turn call our `resumeWith` at a later point in time.

Reply:
I found this non braced `if-else` code in Scala 3 very hard to read 😅 I guess I still need to get used to it. Perhaps using `if` statements it can be a bit better readable?

Reply:
> Perhaps using if statements it can be a bit better readable?

you mean like this 
```
    if (result == Continuation.State.Resumed) return Continuation.State.Suspended
    if (result.isInstanceOf[Left[_, _]]) throw result.asInstanceOf[Left[Throwable, _]].value
    if (result.isInstanceOf[Right[_, _]]) return result.asInstanceOf[Right[_, _]].value
```

to be honest in this case I find the use of `return` a bit weird in Scala (more than the `if-else`) 😂 
but happy to change it once confirmed this is what you meant

Reply:
>I also wonder if Either is the way to go here, or we should use union types here instead since we never rely on the bias. Not sure if union types are free of the allocation overhead. 

I think union types are unboxed but it looks like the choice of `Either` was done intentionally to distinguish the error from the values like [override def resume(result: Either[Throwable, Any | Null])](https://github.com/47deg/TBD/blob/23eaf30ce4a817ab457ff5b08888263d2174e520/continuationsPlugin/src/main/scala/continuations/jvm/internal/ContinuationImpl.scala#L11).

Reply:
We can investigate it later with benchmarks, but this is being used in a hot-loop. Kotlin has a special type Result that avoids allocation in the non-Throwable case and pays a penalty for Throwable where it’s arguably neglectable.

But let’s post-pone that. It’s not important for functionality, and doing performance optimisations make more sense when backed by benchmarks.

Reply:
> > Perhaps using if statements it can be a bit better readable?
> 
> you mean like this?
> ```
>     if (result == Continuation.State.Resumed) return Continuation.State.Suspended
>     if (result.isInstanceOf[Left[_, _]]) throw result.asInstanceOf[Left[Throwable, _]].value
>     if (result.isInstanceOf[Right[_, _]]) return result.asInstanceOf[Right[_, _]].value
>     result
> ```
> 
> to be honest in this case I find the use of `return` in Scala a bit weird  (more than the `if-else`) 😂 
> but happy to change it once confirmed this is what you meant

I think he just meant

```scala
If(){
  ...
} else {
  ...
} ...
```

Reply:
Yes, that is exactly what I meant @jackcviers 👍

Reply:
added the braces in [1a5f9ef](https://github.com/47deg/TBD/pull/52/commits/1a5f9ef8011ed082a480c0da6afa4bae1cc71944)